### PR TITLE
APEXMALHAR-2237 : Changes in FSInputModule to support Dynamic partitioning.

### PR DESCRIPTION
**Pull request by yogidevendra**

### CODE REVIEW

**Comment by yogidevendra** [comment]
@DT-Priyanka Could you please review this?

**Comment by DT-Priyanka** [comment]
I am not aware why this change is made, can you please explain. We have seen some issues with dynamic partitioning and hence decided to go with static. Why are we changing behavior again?

**Comment by yogidevendra** [comment]
This is based on operator maturity exercise. Dynamic scaling is one of the aspects considered for deciding if given operator/module is mature.

File: apps/filecopy/src/main/resources/META-INF/properties.xml [hunk 1]
```diff
@@ -23,26 +23,10 @@
...
-    <value>128000000</value>
-  </property>
-  <property>
-    <name>dt.operator.HDFSInputModule.prop.scanIntervalMillis</name>
-    <value>10000</value>
-  </property>
-  <property>
-    <name>dt.operator.HDFSInputModule.prop.dedup</name>
-    <value>true</value>
+    <value>hdfs://source-namenode-service:port/path-to-input-directory</value>
```
Comment (DT-Priyanka) [inline apps/filecopy/src/main/resources/META-INF/properties.xml]: I will suggest we should keep default values as before.

File: apps/filecopy/src/main/resources/META-INF/properties.xml [hunk 2]
```diff
@@ -23,26 +23,10 @@
 <configuration>
   <property>
     <name>dt.operator.HDFSInputModule.prop.files</name>
-    <value>hdfs://localhost:54310/user/dtadmin/input</value>
-  </property>
-  <property>
```
Comment (DT-Priyanka) [inline apps/filecopy/src/main/resources/META-INF/properties.xml:28]: why are we removing these properties?

File: apps/filecopy/src/site/conf/cluster-memory-conf.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (DT-Priyanka) [inline apps/filecopy/src/site/conf/cluster-memory-conf.xml]: Why there is new config file for these parameters? Do we follow this practice anywhere else?

File: library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java [hunk 4]
```diff
@@ -220,7 +243,9 @@ public void setBlockSize(long blockSize)
   /**
    * Gets readers count
    * @return readersCount
+   * @Deprecated Use getMinReaders(), getMaxReaders() instead
```
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java]: Add better explanation, it's not clear that not having this configuration and setting others mean that we are using dynamic partitioning. Also why two parameters now, is one more question I get when I read this javadoc.

File: library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java [hunk 5]
```diff
@@ -108,8 +114,25 @@ public void populateDAG(DAG dag, Configuration conf)
     }
 
     blockReader.setBasePath(files);
-    if (readersCount != 0) {
-      dag.setAttribute(blockReader, Context.OperatorContext.PARTITIONER, new StatelessPartitioner<FSSliceReader>(readersCount));
+    
+    if (readersCount != 0 && minReaders == 0) {
+      minReaders = readersCount;
```
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java]: remove extra line above.

File: library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java [hunk 6]
```diff
@@ -276,7 +303,61 @@ public int getBlocksThreshold()
...
+  public long getRepartitionCheckInterval()
+  {
+    return repartitionCheckInterval;
+  }
+  
+  /**
+   * Sets Interval for re-evaluating dynamic partitioning
+   * @param repartitionCheckInterval interval for re-evaluating dynamic partitioning
+   */
+  public void setRepartitionCheckInterval(long repartitionCheckInterval)
```
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java:323]: can we call it as repartitionInterval ?

**Review by DT-Priyanka — COMMENTED** [review]
In general, can we add any tests to verify these new parameters and behavior?

File: apps/filecopy/src/site/conf/cluster-memory-conf.xml (hunk 3 above)
Comment (yogidevendra) [reply apps/filecopy/src/site/conf/cluster-memory-conf.xml]: Single configuration is not suitable for all environments. So, we are trying to capture them separately. 

properties.xml would capture mandatory properties for which there cannot be system default.
Other properties which are advanced but can be initialized with meaningful defaults are captured under apconf. These can be used using -apconf from apex cli.

File: library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java (hunk 4 above)
Comment (yogidevendra) [reply library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java]: I retained older parameter for maintaining backward compatibility. Other PR #439  would mark this class as evolving. Then we can remove the older parameter.

File: library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java (hunk 6 above)
Comment (yogidevendra) [reply library/src/main/java/com/datatorrent/lib/io/fs/FSInputModule.java:323]: I thought so, but then I realized that interval is just to check if repartition is needed. Repartition may not always happen after this interval.

File: apps/filecopy/src/main/resources/META-INF/properties.xml (hunk 1 above)
Comment (yogidevendra) [reply apps/filecopy/src/main/resources/META-INF/properties.xml]: Default values are changed to reflect the expected user input (no need to refer to documentation for specifying the value). Anyways, path given in earlier default value may or may not exists on the end user system.

File: apps/filecopy/src/site/conf/cluster-memory-conf.xml (hunk 3 above)
Comment (yogidevendra) [reply apps/filecopy/src/site/conf/cluster-memory-conf.xml]: These are not new parameters. Existing properties in the operators are exposed with getters, setters.

**Comment by yogidevendra** [comment]
@DT-Priyanka All review comments are answered. Could you please have a look.

File: apps/filecopy/src/main/resources/META-INF/properties.xml (hunk 1 above)
Comment (DT-Priyanka) [reply apps/filecopy/src/main/resources/META-INF/properties.xml]: Path may not exist. But localhost and default port is likely to exist. I don't have firm opinion about this, we can go with what looks right to majority.

File: apps/filecopy/src/site/conf/cluster-memory-conf.xml (hunk 3 above)
Comment (DT-Priyanka) [reply apps/filecopy/src/site/conf/cluster-memory-conf.xml]: So if someone wants to run this app, he will have to create new config file consolidating all these properties or there is a way to provide multiple properties file to one application?

**Review by DT-Priyanka — COMMENTED** [review]
Rest of the changes looks okay, will merge it if there are no further comments by tomorrow.

File: apps/filecopy/src/site/conf/cluster-memory-conf.xml (hunk 3 above)
Comment (yogidevendra) [reply apps/filecopy/src/site/conf/cluster-memory-conf.xml]: launch with -apconf option should address this.

**Pull request merged**